lib/posix/pthread_mutex.c

```diff
@@ -9,6 +9,18 @@
 #include <wait_q.h>
 #include <posix/pthread.h>
 
+/* Memory pool for pthread space */
+K_MEM_POOL_DEFINE(posix_mutex_sem, sizeof(struct k_sem),
+		  sizeof(struct k_sem), 32, 4);
```

Comment:
Shouldn't this use a `k_mutex` instead of a `k_sem`? There are some subtle differences that will make it harder to implement by using `k_sem`.  For instance, depending on the mutex type, threads may lock mutexes if they have been already locked.  Please see the man page for `pthread_mutex_lock()` for a table with the behavior for each type:

```
                   ┌───────────┬────────────┬────────────────┬───────────────────────┐
                   │Mutex Type │ Robustness │     Relock     │ Unlock When Not Owner │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │NORMAL     │ non-robust │ deadlock       │ undefined behavior    │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │NORMAL     │ robust     │ deadlock       │ error returned        │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │ERRORCHECK │ either     │ error returned │ error returned        │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │RECURSIVE  │ either     │ recursive      │ error returned        │
                   │           │            │ (see below)    │                       │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │DEFAULT    │ non-robust │ undefined      │ undefined behavior†   │
                   │           │            │ behavior†      │                       │
                   ├───────────┼────────────┼────────────────┼───────────────────────┤
                   │DEFAULT    │ robust     │ undefined      │ error returned        │
                   │           │            │ behavior†      │                       │
                   └───────────┴────────────┴────────────────┴───────────────────────┘

       †     If the mutex type is PTHREAD_MUTEX_DEFAULT, the behavior of pthread_mutex_lock() may
             correspond to one of the three other standard mutex types as described in the  table
             above.  If  it  does not correspond to one of those three, the behavior is undefined
             for the cases marked †.

```

Reply:
No. I have forwarded you a mail which we discussed with Andy Ross for this.

Reply:
please use a sys_mem_pool instead of a k_mem_pool